laurent22/joplin
# Mobile: Fixes #10409: Fix accepting encrypted shared notebooks

# Summary

This pull request adds support for decrypting data longer than 256 bytes in [RSA.react-native.ts](https://github.com/laurent22/joplin/compare/dev...personalizedrefrigerator:pr/mobile/fix-accept-encrypted-shared-notebook?expand=1#diff-d2038dbede8167cd4b88193c6c58c5dab335a0c9d0118178a57a3e35c25af52b). 

Previously, [ppkEncryptionHandler](https://github.com/laurent22/joplin/blob/90ec1f5bc6d421fb1361dd864f088b35a21fdc67/packages/lib/services/e2ee/ppk.ts#L105) was called with data longer than 256 bytes while accepting a share, causing an error.

This pull request **does not** add support for long-data encryption mobile. Long-data encryption should only be necessary for sharing encrypted notebooks.

# Notes

- [`node-rsa`](https://github.com/rzcoder/node-rsa/blob/e7e7f7d2942a3bac1d2e132a881e5a3aceda10a1/src/libs/rsa.js#L227) seems to use [ECB-mode](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Electronic_codebook_(ECB)) (breaking the data into chunks, then encrypting/decrypting each individually) for encrypting/decrypting multiple blocks.
- The key size is now included in the `RSAKeyPair` data structure. This is used to calculate the block size. While it's possible to get the key size from the public key, [it seems to be more complicated](https://github.com/laurent22/joplin/commit/3b6c53ca6357c0ca3100d1580e139145d1a5d8b6#diff-d2038dbede8167cd4b88193c6c58c5dab335a0c9d0118178a57a3e35c25af52bR33) than providing a copy of the key size to `encrypt`/decrypt`.
- Using a short key with longer data with length that's multiple of 128 bytes seems to fail. We currently use a key size of 256 bytes and is tested for a key size of 256 above.

# Testing plan

This pull request extends the existing automated tests that run on app startup. It can also be tested manually by:
1. Setup (if not already done):
   1. Set up two-user sync with Joplin server, **where both users have encryption enabled** (if not already set-up).
   2. Sync a mobile client with one account, a desktop client with the other.
2. Create a new notebook from the desktop app.
3. Share that notebook with the mobile client user.
4. Sync the mobile client and accept the share.
5. Sync the desktop client and restart the mobile client.
   - This seems to be necessary on iOS and seems to be a separate issue from #10409.
6. Create a new note in the shared notebook from the mobile client and attach a drawing.
7. Sync the desktop client and verify that the just-created note is visible.
8. Edit the note on desktop and sync both clients.
9. Verify that the desktop changes can be seen on mobile.

This has been tested successfully on an iOS 17.4 simulator and an Android 14 emulator.
